scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/TerminationData.java

```diff
@@ -153,23 +156,62 @@ boolean taskTerminated(JobId j, String taskName) {
     }
 
     void handleTermination(final SchedulingService service) throws IOException, ClassNotFoundException {
-        for (TaskTerminationData taskToTerminate : tasksToTerminate.values()) {
-            RunningTaskData taskData = taskToTerminate.taskData;
-            if (taskToTerminate.terminatedWhileRunning()) {
-                terminateRunningTask(service, taskToTerminate, taskData);
-            }
+
+        terminateTasks(service);
+
+        restartWaitingTasks(service);
+
+        terminateJobs(service);
+    }
+
+    private void terminateJobs(final SchedulingService service) {
+        for (JobId jobId : jobsToTerminate) {
+            service.terminateJobHandling(jobId);
         }
+    }
 
+    private void restartWaitingTasks(final SchedulingService service) {
         for (final TaskRestartData restartData : tasksToRestart.values()) {
             service.getInfrastructure().schedule(new Runnable() {
                 public void run() {
                     service.getJobs().restartWaitingTask(restartData.taskId);
                 }
             }, restartData.waitTime);
         }
+    }
 
-        for (JobId jobId : jobsToTerminate) {
-            service.terminateJobHandling(jobId);
+    private void terminateTasks(final SchedulingService service) {
+
+        if (tasksToTerminate.values().isEmpty()) {
+            return;
+        }
+
+        ExecutorService parallelTerminationService = Executors.newFixedThreadPool(tasksToTerminate.values().size());
```

Comment:
The executor service should be shutdown at the end of this method, otherwise it will leave alive threads